pkg/model/utxo/dust_db.go

```diff
@@ -0,0 +1,223 @@
+	}
+
+	if dustAllowanceBalance == 0 {
+		// Remove from database
+		return mutations.Delete(addressBytes)
+	} else {
+		return mutations.Set(addressBytes, bytesFromDust(dustAllowanceBalance, dustOutputCount))
+	}
+
+	return nil
+}
+
+func (u *Manager) applyDustDiff(dustDiff map[string]*DustDiff) error {
+
+	mutations := u.dustStorage.Batched()
+	for addr, diff := range dustDiff {
+		if err := u.applyDustDiffForAddress(addr, diff.DustAllowanceBalanceDiff, diff.DustOutputCount, mutations); err != nil {
+			mutations.Cancel()
+			return err
+		}
+	}
+	return mutations.Commit()
+}
+
+func (u *Manager) applyDustDiffForAddress(addressString string, dustAllowanceBalanceDiff int64, dustOutputCountDiff int64, mutations kvstore.BatchedMutations) error {
+
+	addressBytes := []byte(addressString)
+
+	dustAllowanceBalance, dustOutputCount, err := u.readDustForAddress(addressBytes)
+	if err != nil {
+		return err
+	}
+
+	newDustAllowanceBalance := int64(dustAllowanceBalance) + dustAllowanceBalanceDiff
+	newDustOutputCount := dustOutputCount + dustOutputCountDiff
+
+	if newDustOutputCount < 0 || newDustAllowanceBalance < 0 {
+		// Count or balance cannot be negative
+		return fmt.Errorf("%w: %s dustAllowanceBalance %d, dustOutputCount %d", ErrInvalidDustForAddress, hex.EncodeToString(addressBytes), dustAllowanceBalance, dustOutputCount)
+	}
+
+	return u.storeDustForAddress(addressBytes, uint64(newDustAllowanceBalance), newDustOutputCount, mutations)
+}
+
+func (u *Manager) applyNewDustWithoutLocking(newOutputs Outputs, newSpents Spents) error {
+	dustDiff := make(map[string]*DustDiff)
+
+	for _, output := range newOutputs {
+		// Add new dust
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff += int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
+					diff.DustOutputCount += 1
+				} else {
+					dustDiff[address] = NewDustDiff(0, 1)
+				}
+			}
+		}
+	}
+
+	for _, spent := range newSpents {
+
+		// Remove spent dust
+		output := spent.Output()
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff -= int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(-int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
```

Comment:
you should convert the address bytes to string, but the String function uses hex encoding